### DISCUSSION TITLE
Possible to define store in library and consume in app?

### DISCUSSION THREAD

**Discussion by aduarterengifo** [pmndrs/zustand#2188] [Q&A]
Hello!

I'm in a monorepo, 

I would like to move my zustand stores to a separate package that is consumed by many applications. 

I'm wondering if this is possible?

I've tried it with no luck. 

my stores are always empty.

### stores library:
```json
{
  "name": "stores",
  "peerDependencies": {
    "react": "^18.2.0",
    "zustand": "^4.4.1"
  },
}
```

```ts
// index.ts

import { create } from "zustand";

type store = {
  color: "red" | "blue",
  setColor: (color: "red" | "blue") => void
};

export const useStore = create<store>((set) => ({
  color: "blue"
  setColor: (color) => set({ color}),
}));
```

### app A
```json
{
  "name": "appA",
  "dependencies": {
    "next": "14.0.0",
    "react": "^18.2.0",
    "zustand": "^4.4.1",
    "stores": "workspace:*",
  },
}
```

```tsx 
// app/layout.tsx

import { useStore } from "store";

export default async function  Page() {
  useStore.setState({
    color: "red",
  });
  ...
}
```

component B:
```tsx 
// app/color/page.tsx

import { useStore } from "store";

export default async function  Page() {
  return (
    <div>{ useStore.getState().color}</div>
  )
}
```

*Also not sure of this is relevant but these are Next.js 13, app dir, apps. and this store is meant for server components.

**Comment by somewonderfulguy** [comment]
@lofsigma
You can share stores between your apps but you need to understand that importing store using monorepo you will get different instances of store in your app A and app B. Meaning, `useStoreA !== useStoreB`, e.g. `{} !== {}`.

You need to pass store. Please, pay attention to this page in Zustand documentation: [Initialize state with props](https://zustand.docs.pmnd.rs/learn/guides/initialize-state-with-props).

How I made it. I have main app and sub-app. Simplified example.

### Main App
```typescript
import SubApp from '@repo/sub-app'
/* importing type from packages because sub-app cannot import anything from main app - it will cause circular dependency error */
import { BearStore } from '@repo/shared/types'
import { create } from 'zustand'

const useBearStore = create<BearStore>()((set) => ({
  bears: 0,
  increase: (by) => set((state) => ({ bears: state.bears + by })),
  clear: () => set({ bears: 0 })
}));

const App = () => {
  const bears = useBearStore((s) => s.bears)
  const increase = useBearStore((s) => s.increase)
  
  return (
    <>
      <div>bears: {bears}</div>
      <div>
        <button type="button" onClick={() => increase(1)}>Add bear</button>
      </div>
      <Sub useBearStore={useBearStore} />
    </>
  )
}
```

### Sub App
```typescript
import { BearStore } from '@repo/shared/types'

type Props = {
  useBearStore: UseBoundStore<StoreApi<BearStore>>
}

const SubApp = ({ useBearStore }: Props) => {
  /* here, you can actually create React context provider (with this store) and custom hook to access context, so deep down in
  * components tree you can access `useBearStore` using your custom hook.
  * For performant context you can see this video:
  * [Making React Context FAST!](https://youtu.be/ZKlXqrcBx88?si=NHHqu6Svcnt2uqXf)
  */
  const clearBears = useBearStore((s) => s.clear)

  return (
    <button type="button" onClick={clearBears}>
      Clear bears
    </button>
  )
}
```

### Module Federation
Module federation allows to share the same instance. So, you can avoid passing store as prop and adding such as complexity as putting store in context. The only disadvantage of Module Federation is you need to deploy it separately. However, deploying store make sense to share the same instance(s) of store(s).
Here, you can see examples: [State Management for Module Federation Four Ways](https://youtu.be/njXeMeAu4Sg?si=JtZcjgyYENKXhTI9).

---

UPD: I tried module federation for moving state to separate app that is federated module and expected to be deployed separately (not tried yet). It works nicely. And the instance is the same - you don't have to pass store object as prop. I keep this as separate package inside monorepo because I want to access TS types. Yes, you'll need to specify types explicitly when using module federation as it exposes simple JS. If you know nothing about Module Federation, here's another video [Vite and Module Federation Makes Micro-Frontends EASY!](https://youtu.be/t-nchkL9yIg?si=pOGbdvGljgVR4sPQ). On the channel you'll find a lot videos related to this topic.
